zebrad/src/components/inbound.rs

```diff
@@ -110,96 +131,131 @@ impl Service<zn::Request> for Inbound {
+                    Ok((outbound, address_book)) => {
+                        let downloads = Box::pin(Downloads::new(
+                            Timeout::new(outbound, BLOCK_DOWNLOAD_TIMEOUT),
+                            Timeout::new(verifier, BLOCK_VERIFY_TIMEOUT),
+                            self.state.clone(),
+                        ));
+                        self.network = Setup::Initialized {
+                            address_book,
+                            downloads,
+                        };
+                    }
+                    Err(TryRecvError::Empty) => {
+                        // There's no setup data yet, so keep waiting for it
+                        self.network = Setup::AwaitingNetwork {
+                            network_setup,
+                            verifier,
+                        };
+                    }
+                    Err(error @ TryRecvError::Closed) => {
+                        // Mark the service as failed, because network setup failed
+                        error!(?error, "inbound network setup failed");
+                        self.network = Setup::FailedRecv {
+                            failed_setup: network_setup,
+                        };
+                        return Poll::Ready(Err(error.into()));
+                    }
                 }
-                Err(e @ TryRecvError::Closed) => {
-                    // In this case, report that the service failed, and put the
-                    // failed oneshot back so we'll fail again in case
-                    // poll_ready is called after failure.
-                    self.network_setup = Some(rx);
-                    return Poll::Ready(Err(e.into()));
-                }
-            };
+            }
         }
 
-        // Clean up completed download tasks
-        if let Some(downloads) = self.downloads.as_mut() {
-            while let Poll::Ready(Some(_)) = downloads.as_mut().poll_next(cx) {}
+        // Unfortunately, we can't combine these matches into an exhaustive match statement,
+        // because they use mutable references, or they depend on the state we've just modified.
+
+        // Make sure we left the network setup in a valid state
+        if matches!(self.network, Setup::FailedInit) {
+            unreachable!("incomplete Inbound initialization");
+        }
+
+        // If network setup failed, report service failure
+        if let Setup::FailedRecv { failed_setup } = &mut self.network {
+            // TryRecvError is not cloneable, so we have to generate a new error from the oneshot,
+            // rather than re-using a clone of the original error
+            let failed_response = failed_setup.try_recv();
+            if let Err(error @ TryRecvError::Closed) = failed_response {
+                return Poll::Ready(Err(error.into()));
+            } else {
+                unreachable!("unexpected response from failed Inbound network setup oneshot");
+            }
```

Comment:
we should probably still log the error in this panic, so we know what unexpected error condition we actually encountered.

Reply:
```suggestion
            } else {
                unreachable!("unexpected response from failed Inbound network setup oneshot: {:?}", failed_response);
            }
```